Kernel/FileSystem/Ext2Base.h

```diff
@@ -0,0 +1,297 @@
+    u32 journal_inode_num;
+    u32 journal_device;
+    u32 last_orphan;
+    u32 htree_hash_seed[4];
+    u8 default_hash_version;
+    u8 journal_backup_type;
+    u16 group_descriptor_size_num;
+    u32 default_mount_options;
+    u32 first_meta_block_group;
+    u32 fs_create_time;
+    u32 journal_inode_backup[17];
+    u32 num_blocks_high;
+    u32 num_reserved_blocks_high;
+    u32 num_free_blocks_high;
+    u16 min_extra_inode_size;
+    u16 want_extra_inode_size;
+    u32 misc_flags;
+    u16 raid_stride;
+    u16 mmp_interval;
+    u64 mmp_block;
+    u32 raid_stripe_width;
+    u8 log_groups_per_flex;
+    u8 reserved_char_pad;
+    u16 reserved_pad;  // Padding to next 32bits
+    u32 reserved[162]; // Padding to the end of the block
+
+    InodeIndex first_inode() const
+    {
+        if (rev_level == to_underlying(Revision::Original))
+            return to_underlying(Ext2::InodeNums::FirstInode);
+        else
+            return first_inode_num;
+    }
+    size_t inode_size() const
+    {
+        if (rev_level == to_underlying(Revision::Original))
+            return good_old_inode_size;
+        else
+            return inode_size_num;
+    }
+    size_t block_size() const
+    {
+        return (1 << min_block_log_size) << block_log_size;
+    }
+    size_t fragment_size() const
+    {
+        return (1 << min_fragment_log_size) << fragment_log_size;
+    }
+    // Returns the exponent of the block's size.
+    size_t block_size_bits() const
+    {
+        return block_log_size + min_block_log_size;
+    }
+    size_t inodes_per_block() const
+    {
+        return block_size() / inode_size();
+    }
+    size_t fragments_per_block() const
+    {
+        return block_size() / fragment_size();
+    }
+    size_t addresses_per_block() const
+    {
+        return block_size() / sizeof(u32);
+    }
+    size_t group_descriptor_size() const
+    {
+        if (incompatible_features & ext4_feature_incompat_64bit)
+            return group_descriptor_size_num;
+        else
+            return min_group_descriptor_size;
+    }
+    size_t group_descriptors_per_block() const
+    {
+        return block_size() / group_descriptor_size();
+    }
+};
+
+// Ensure the ext2 super block is 1024 bytes in size.
+static_assert(sizeof(SuperBlock) == 1024);
+
+// Ext2 directory file types. Only the low 3 bits are used. The
+// other bits are reserved for now.
+enum class FileType : u8 {
+    Unknown,   // Unknown file type
+    RegFile,   // Regular old file
+    Directory, // Directory
+    CharDev,   // Character device
+    BlockDev,  // Block device
+    FIFO,      // FIFO
+    Socket,    // Socket
+    SymLink,   // Symbolic link
+    Max        // The maximum file type plus one.
+};
+
+// The new version of the directory entry. Since EXT2 structures are
+// stored in intel byte order, and the name_len field could never be
+// bigger than 255 chars, it's safe to reclaim the extra byte for the
+// file_type field.
+
+struct DirEntry {
```

Comment:
No abbreviations as per the coding style guidelines

Reply:
Wouldn't it be more canonical for it to be DirEntry rather than DirectoryEntry? I thought the coding style had an exemption for shorthands if it's more canonical that way

> Use full words, except in the rare case where an abbreviation would be more canonical and easier to understand.

Reply:
"dir" is by no means more canonical than "directory".